### ISSUE TITLE
RFC: Heterogenous Unions / Sum Types

### ISSUE THREAD

**Issue by ohamel-softwaresecure** [souffle-lang/souffle#1395]
Due to time constraints, the following is copy/pasted from PR #1310, which provides an implementation of sum-types.  First is @b-scholz's comment regarding their ideal, followed by my request for clarification/disordered thoughts.

@b-scholz:

We had more discussion at USYD about ADTs. We feel that for the sake of consistency, union types are the right way forward as a first step. The rationale is that we have introduced union types for primitive types some time ago and it might be confusing that we have different typing strategies.

Said this, we feel we can simulate ADTs with union types having a small little tweak. As discussed, we have the following problem:
```
.type X = A | B 
.type A = [...]
.type B = [...]
```
Type X would have a purely syntactic typing behavior in our system. To make it dynamic as perhaps needed for ADT translation, we would need to box type X.
```
.type X = [box: A | B] 
.type A = [...]
.type B = [...]
```
If we implement it well, simple boxing (over a single field) may be implemented very efficiently (i.e. without a double lookup).

Records in a union class are mapped to a single record (as you demonstrate for ADTs and as it is implemented in our type-system branch from last year). For simple boxed types, we can carry some additional flags in the type identifier field.

After implementing polymorphic union types, we are quite open-minded to implement ADTs if the need arises. For us, this is only a question of prioritization.

@ohamel-softwaresecure:
**TL;DR:** 
I'm confused as to what syntax and semantics you're looking for. We may need to write some toy programs to provide concrete examples.

Gist of the problem: How do unions of records work?

**Minimally Edited Stream of Consciousness Dump:**
I've some questions regarding the semantics of heterogeneous unions but first, to ensure I understood correctly:

**Summary of Changes:**
1) Core: Add support for structurally-heterogeneous unions via the same tagged-box mechanism as the PR.
2) Ergonomics: Automatically box/unbox as required. (AST transform)
3) Ergonomics: Filter branches when unifying with another union.
4) Optimization: Elide boxing when union branches are subtypes of the same type. (e.g. numeric type)
5) Optimization: Flatten boxed single-element non-recursive records. (Extra, needed for below.)

Given the above, keeping the proposed ADT syntax/behaviour above requires only the syntax changes + desugaring each branch into a 1-record.

**Questions:**
Consider the following type declarations:
```
// two disjoint subsets of `number`
.type S_a <: number
.type S_b <: number

// nominally distinct, structurally equiv
.type R2_a  = [ x: string, y: number ]
.type R2_b  = [ x: string, y: number ]
// and a structurally distinct record
.type R1    = [ x: S_a ]

//  NEW: hypothetical generalisation of disjoint subtypes to any type
.type R2_a_a <: R2_a
.type R2_a_b <: R2_a
.type R2_a_b_a <: R2_a_b // subtype of subtype?

.type USub_S  = S_a | S_b         // union of non-record primitive subtype
.type USub_R  = R2_a_a | R2_a_b   // NEW: union of generalised subtypes
.type UHet_R2 = R2_a | R2_b       // NEW: heterogenous union of structurally equiv types
.type UHet_R1 = R2_a | R1         // NEW: heterogenous union of structurally equiv types
// NEW: Kitchen sink
.type USink   = S_a | S_b | R2_a | R2_b | R1
```

Which of these are valid? What are their semantics? In what cases are we allowed to discern which type(s) a value with a union type satisfied?

Unions currently operate under type-erasure: once a value is stuffed into a location with a union type you cannot find out which type(s) it satisfied. This is fine so long as the union is homogeneous (every type in the union extend the same 'root' type, i.e. the same record type, `number`, `symbol`, etc), but as soon as the union is heterogeneous then it effectively becomes an opaque type and the value can only be used for equality testing. (Assuming we change the run-time to ensure all distinct.)

The proposal text above seems to suggest that top-level unions would use erasure, while unions-in-record-decls (new syntax) would box-and-tag. (`.type X = [box: A | B]`) This is rather unsatisfying for various reasons:
1) For cases where the branches are records, the syntax to match on a branch isn't so bad: `x = [ A [ a1, a2, ... ] ]`.
2) For non-records, it's unclear and potentially weird: `x = [ symbol y ]`.
3) It's surprising and confusing why I have to put my union inline in a record if I want to be able to type-match later on.
4) If we allow top-level heterogeneous unions then it isn't clear when unions are type-erased and when they're not. (Make them all heterogeneous and add an optimization that detects which can be erased?)
5) If we don't allow top-level heterogeneous unions then we're basically back to the PR as is except  now there are heterogeneous union types that are distinct from erasure unions, can only be expressed within record decls, and must be typed out in full for every use (you can't give them names b/c `.type A = B | C` is an erasure/homogeneous union).
